cpp/src/slice2cs/Gen.cpp

```diff
@@ -1056,75 +1043,126 @@ Slice::Gen::TypesVisitor::visitClassDefEnd(const ClassDefPtr& p)
     _out << eb;
 }
 
-bool
-Slice::Gen::TypesVisitor::visitInterfaceDefStart(const InterfaceDefPtr& p)
+void
+Slice::Gen::TypesVisitor::visitSequence(const SequencePtr& p)
 {
     string ns = getNamespace(p);
+    string name = p->mappedName();
 
+    string typeS = typeToString(p, ns);
     _out << sp;
-    emitAttributes(p);
+    _out << nl << "public sealed class " << name << "Helper";
+    _out << sb;
 
-    _out << nl << "[Ice.SliceTypeId(\"" << p->scoped() << "\")]";
-    _out << nl << "public partial interface " << p->mappedName();
+    _out << sp << nl << "public static void write(Ice.OutputStream ostr, " << typeS << " v)";
+    _out << sb;
+    writeSequenceMarshalUnmarshalCode(_out, p, ns, "v", true, false);
+    _out << eb;
 
-    list<string> baseNames;
-    for (const auto& q : p->bases())
-    {
-        baseNames.push_back(getUnqualified(q, ns));
-    }
+    _out << sp << nl << "public static " << typeS << " read(Ice.InputStream istr)";
+    _out << sb;
+    _out << nl << typeS << " v;";
+    writeSequenceMarshalUnmarshalCode(_out, p, ns, "v", false, false);
+    _out << nl << "return v;";
+    _out << eb;
+    _out << eb;
 
-    if (baseNames.empty())
+    if (auto metadata = p->getMetadataArgs("cs:generic"))
     {
-        baseNames.emplace_back("Ice.Object");
-    }
+        string_view type = *metadata;
+        if (type == "List" || type == "LinkedList" || type == "Queue" || type == "Stack")
+        {
+            return;
+        }
 
-    _out << " : ";
-    bool emitSep = false;
-    for (const auto& baseName : baseNames)
-    {
-        if (emitSep)
+        if (!p->type()->isClassType())
         {
-            _out << ", ";
+            return;
         }
-        emitSep = true;
-        _out << baseName;
-    }
 
-    _out << sb;
-    return true;
+        //
+        // The sequence is a custom sequence with elements of class type.
+        // Emit a dummy class that causes a compile-time error if the
+        // custom sequence type does not implement an indexer.
+        //
+        _out << sp;
```

Comment:
I assume this was already there, but seems and odd test. If a sequence doesn't implement the required methods the code should already fail to compile.

Reply:
We need to review this code in a follow-up PR.